src/commands/registerWorkspaceCommand.ts

```diff
@@ -0,0 +1,104 @@
+                    learnMoreLink = 'https://aka.ms/AA5y2rd';
+                    break;
+                case RemoteKind.wsl:
+                    message = 'This operation is not supported because the Docker extension is currently running outside of WSL.';
+                    switchTitle = 'Switch to WSL';
+                    learnMoreLink = 'https://aka.ms/AA5xvjn';
+                    break;
+                case RemoteKind.devContainer:
+                    message = 'This operation is not supported because the Docker extension is currently running outside of your container.';
+                    switchTitle = 'Switch to Container';
+                    learnMoreLink = 'https://aka.ms/AA5xva6';
+                    break;
+                default:
+                    // Assume this works rather than block users on unknown remotes
+                    return;
+            }
+
+            context.telemetry.properties.cancelStep = 'switchExtensionKind';
+            const switchBtn: MessageItem = { title: switchTitle };
+            await ext.ui.showWarningMessage(message, { learnMoreLink }, switchBtn);
+            updateExtensionKind('workspace');
+
+            context.telemetry.properties.cancelStep = 'requiresReload';
+            let reloadMessage: string = 'This change to the Docker extension requires reloading VS Code to take effect.';
+            let reload: MessageItem = { title: 'Reload Now' };
+            await ext.ui.showWarningMessage(reloadMessage, reload);
+
+            // Add a one-off event here before reloading the window otherwise we'll lose telemetry for this code path
+            await callWithTelemetryAndErrorHandling('verifyIsWorkspaceExtension', (newContext: IActionContext) => {
+                Object.assign(newContext, context);
+            });
+
+            await commands.executeCommand('workbench.action.reloadWindow');
+
+            context.telemetry.properties.cancelStep = 'reloading';
+            // throw an exception just to make sure we don't try to continue the command before the window is fully reloaded
+            throw new UserCancelledError();
+        }
+    }
+}
+
+function updateExtensionKind(newKind: string): void {
+    const settingKey: string = 'remote.extensionKind';
+    const config: WorkspaceConfiguration = workspace.getConfiguration();
+    const values = config.inspect(settingKey);
+    let target: ConfigurationTarget;
+    let value: {};
+
+    // If the setting is already defined as a workspace setting - overwrite that
+    if (typeof values.workspaceValue === 'object' && values.workspaceValue !== null && values.workspaceValue[extensionId]) {
+        target = ConfigurationTarget.Workspace;
+        value = values.workspaceValue;
+    } else { // otherwise update the global setting
+        target = ConfigurationTarget.Global;
+        if (typeof values.globalValue === 'object' && values.globalValue !== null) {
+            value = values.globalValue || {};
```

Comment:
nit: The ||{} is just defensive, right?  If it's an object and not null, then can it be falsy?

Reply:
You're right - shouldn't be possible to be falsy